### PR TITLE
Setup basic logging

### PR DESCRIPTION
**Pull request by torkleyy**
TODO: decide on a simple logging implementation and initialize it in the `Application`

---
This change is [Reviewable](https://reviewable.io/reviews/amethyst/amethyst/513)

### CODE REVIEW

**Comment by Rhuagh** [comment]
Reviewed 20 of 20 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/amethyst/amethyst/513)*

**Comment by Xaeroxe** [comment]
I like this a lot!  Thank you!

Question for @amethyst/engine-devs and @Rhuagh though, A lot of our crates are starting to get these common dependencies, such as `log` and `shrev`.  Perhaps we should identify our "framework dependencies" and re-export them in core just to make maintenance easier.

**Comment by torkleyy** [comment]
We have to be careful though because if we do this with private dependencies (which log is) we would have to make a breaking change when we upgrade them.

**Comment by torkleyy** [comment]
So..any suggestions for the logger implementation?

**Comment by Xaeroxe** [comment]
I'm content with the idea of having to make breaking changes when upgrading log, chances are pretty good users are going to want to log their own things too.  (I know I've done this for a few games I've worked on)

**Comment by torkleyy** [comment]
Ah, nicely observed @Xaeroxe. I think it's okay to expose the `log` crate then :+1:

**Comment by Emilgardis** [comment]
Related
#124 
#460

**Comment by Rhuagh** [comment]
`simple_logger` for examples imho.

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/amethyst/amethyst/513#-:-L1CFzavHzYodD5i-q1M:bf16tt1)*

**Comment by torkleyy** [comment]
Only problem with that is that the last commit was two years ago :( It looks pretty much like `env_logger`, so we could as well just use that.

**Comment by Rhuagh** [comment]
That or `simplelog`, I don't mind either. For production games it will likely be log4rs or similar

**Comment by Rhuagh** [comment]
`fern` looks nice imho

**Comment by Rhuagh** [comment]
Use `env_logger` for the examples. How far is this from completion? 

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/amethyst/amethyst/513#-:-L1wrrOUfkht-nK5H2Sj:bp0fouh)*

**Comment by torkleyy** [comment]
This is pretty easy to complete, and I planned to use fern. Just didn't have time for it yet.

**Comment by Rhuagh** [comment]
Ok. Use `fern`, I like that one the best :)

**Comment by Xaeroxe** [comment]
Bad robot

**Comment by torkleyy** [comment]
Example output:

```
[amethyst::app][INFO] Initializing Amethyst...
[amethyst::app][INFO] Version: v0.5.1-398-g40fa47b
[amethyst::app][INFO] Platform: x86_64-unknown-linux-gnu
[amethyst::app][INFO] Amethyst git commit: 40fa47ba289cc87f5fc927a2939e071adff21af0
[amethyst::app][INFO] Rustc version: 1.25.0-nightly Nightly
[amethyst::app][INFO] Rustc git commit: 51b0b3734cbd0ca58c8be3512d53fce2d95f40dd
[amethyst::app][INFO] Engine is shutting down
```

Color may be added later.

**Comment by torkleyy** [comment]
I'd say this is ready for merge; we want to add a logging config later, but we can now finally use logging when creating a PR and don't have to use `println`s and remove them at the end :)

**Comment by torkleyy** [comment]
Atm this will just log to stdout. As I said above, we need to add a logging configuration, but this PR is helpful as is.

**Comment by Emilgardis** [comment]
An easy way to make this configurable is to add one function to `App` and change `App::new`
- `App::new`: `fn new<P: AsRef<Path>>(path: P, initial_state: T) -> Result<Self>`
- - uses `App::with_logger` internally, using the default logger
- `App::with_logger`: `fn _new<P: AsRef<Path>>(path: P, initial_state: T, logger: fern::Dispatch) -> Result<Self>`

**Comment by Rhuagh** [comment]
Reviewed 2 of 6 files at r2, 6 of 6 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[src/app.rs, line 316 at r3](https://reviewable.io:443/reviews/amethyst/amethyst/513#-L2uic9oXGSTQvngxrx1:-L2uic9oXGSTQvngxrx2:bg60s0y) ([raw file](https://github.com/amethyst/amethyst/blob/fdaa31526029d4641e3ef30432dcb00541584084/src/app.rs#L316)):*
> ```Rust
>         use rustc_version_runtime;
> 
>         fern::Dispatch::new()
> ```

Add a `with_logger` function on `ApplicationBuilder`, and use this as the fallback in the `build` function?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/amethyst/amethyst/513)*

File: src/app.rs [hunk 1]
```diff
@@ -151,6 +154,8 @@ impl<'a, 'b> Application<'a, 'b> {
 
     /// Advances the game world by one tick.
     fn advance_frame(&mut self) {
+        trace!("Advancing frame (`Application::advance_frame`)");
```
[the remark this answers is not in the document]
Comment (torkleyy) [reply src/app.rs]: Well you usually filter the logging at this level (if you don't, you shouldn't expect to get only a few lines of output). The whole point of `trace` is to provide information you'll almost never need, but can be quite helpful to localize a bug in very rare cases.

File: examples/sphere/resources/display_config.ron [hunk 2]
```diff
@@ -6,5 +6,5 @@
   multisampling: 1,
   title: "Sphere example",
   visibility: true,
-  vsync: true,
+  vsync: false,
```
[the remark this answers is not in the document]
Comment (torkleyy) [reply examples/sphere/resources/display_config.ron]: I forgot to reset it after running the example (glutin bug currently causes a crash when using vsync).

File: src/app.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Xaeroxe) [reply src/app.rs]: Ok I understand now.  Thanks!

**Comment by Rhuagh** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[src/app.rs, line 316 at r3](https://reviewable.io:443/reviews/amethyst/amethyst/513#-L2uic9oXGSTQvngxrx1:-L3-TAd3YYcdnLC3XWVL:b-9vp0b0) ([raw file](https://github.com/amethyst/amethyst/blob/fdaa31526029d4641e3ef30432dcb00541584084/src/app.rs#L316)):*
<details><summary><i>Previously, Rhuagh (Simon Rönnberg) wrote…</i></summary><blockquote>

Add a `with_logger` function on `ApplicationBuilder`, and use this as the fallback in the `build` function?
</blockquote></details>

Actually, just put this in a `with_default_logger(mut self) -> Self {}` 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/amethyst/amethyst/513)*

**Comment by torkleyy** [comment]
Problem is we don't get all the initialization messages if the logger setup is in a builder method.

**Comment by Rhuagh** [comment]
Hrmph, I guess that's true.

**Comment by Xaeroxe** [comment]
Reviewed 13 of 20 files at r1, 2 of 6 files at r2, 6 of 6 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/amethyst/amethyst/513)*

**Comment by Xaeroxe** [comment]
Overriding

**Comment by Emilgardis** [comment]
This needs to be documented somewhere. Even though it is roughly implemented, making an effort to document it should make it clearer on how to do this properly.

**Pull request merged**